src/analysis/inner_product_space/pi_L2.lean

```diff
@@ -364,13 +460,137 @@ by simp [complex.isometry_of_orthonormal, (dec_trivial : (finset.univ : finset (
 
 open finite_dimensional
 
-/-- Given a natural number `n` equal to the `finrank` of a finite-dimensional inner product space,
-there exists an isometry from the space to `euclidean_space 𝕜 (fin n)`. -/
-def linear_isometry_equiv.of_inner_product_space
-  [finite_dimensional 𝕜 E] {n : ℕ} (hn : finrank 𝕜 E = n) :
-  E ≃ₗᵢ[𝕜] (euclidean_space 𝕜 (fin n)) :=
-((fin_std_orthonormal_basis hn).to_orthonormal_basis
-  (fin_std_orthonormal_basis_orthonormal hn)).repr
+/-! ### Existence of orthonormal basis, etc. -/
+
+section finite_dimensional
+
+variables {v : set E}
+
+variables {A : ι → submodule 𝕜 E}
+
+/-- Given an internal direct sum decomposition of a module `M`, and an orthonormal basis for each
+of the components of the direct sum, the disjoint union of these orthonormal bases is an
+orthonormal basis for `M`. -/
+noncomputable def direct_sum.submodule_is_internal.collected_orthonormal_basis
+  (hV : @orthogonal_family 𝕜 _ _ _ _ (λ i, A i) _ (λ i, (A i).subtypeₗᵢ))
+  [decidable_eq ι] (hV_sum : direct_sum.submodule_is_internal (λ i, A i)) {α : ι → Type*}
+  [Π i, fintype (α i)]
+  (v_family : Π i, orthonormal_basis (α i) 𝕜 (A i)) :
+  orthonormal_basis (Σ i, α i) 𝕜 E :=
+begin
+  apply (hV_sum.collected_basis (λ i, (v_family i).to_basis)).to_orthonormal_basis _,
+  simpa using hV.orthonormal_sigma_orthonormal
+    (show (∀ i, orthonormal 𝕜 (v_family i).to_basis), by simp),
+end
+
+lemma direct_sum.submodule_is_internal.collected_orthonormal_basis_mem [decidable_eq ι]
+  (h : direct_sum.submodule_is_internal A) {α : ι → Type*} [Π i, fintype (α i)]
+  (hV : @orthogonal_family 𝕜 _ _ _ _ (λ i, A i) _ (λ i, (A i).subtypeₗᵢ))
+  (v : Π i, orthonormal_basis (α i) 𝕜 (A i)) (a : Σ i, α i) :
+  h.collected_orthonormal_basis hV v a ∈ A a.1 :=
+by simp [direct_sum.submodule_is_internal.collected_orthonormal_basis]
+
+variables [finite_dimensional 𝕜 E]
+
+-- move this
+lemma _root_.linear_independent.finite {K : Type*} {V : Type*} [division_ring K] [add_comm_group V]
+  [module K V] [finite_dimensional K V] {b : set V} (h : linear_independent K (λ (x:b), (x:V))) :
+  b.finite :=
+cardinal.lt_omega_iff_finite.mp (finite_dimensional.lt_omega_of_linear_independent h)
```

Comment:
Move this! :-)

Reply:
Where should it live?

Reply:
Let's say right after `finite_dimensional.lt_omega_of_linear_independent`?